spring-boot/proxy/src/main/java/eu/coatrack/proxy/security/ApiKeyAuthTokenVerifier.java

```diff
@@ -20,166 +20,124 @@
     }
 
-    private ServiceApi getServiceApiByApiKey(String apiKeyValue) {
+    private String getApiKeyValue(Authentication authentication) {
+        log.debug("Getting API key value from authentication {}.", authentication.getName());
+        Assert.notNull(authentication.getCredentials());
+        Assert.isInstanceOf(String.class, authentication.getCredentials());
+        String apiKeyValue = (String) authentication.getCredentials();
+        Assert.hasText(apiKeyValue);
+        return apiKeyValue;
+    }
 
-        try {
-            log.debug("trying to get service api entity by APIKEY value {}", apiKeyValue);
+    private boolean isAdminsKey(String apiKeyValue) {
+        log.debug("Checking if the API with the value {} is admins API key.", apiKeyValue);
+        return apiKeyValue.equals(ApiKey.API_KEY_FOR_YGG_ADMIN_TO_ACCESS_PROXIES);
+    }
 
-            String urlToGetServiceApi = securityUtil.attachGatewayApiKeyToUrl(
-                    adminBaseUrl + adminResourceToGetServiceByApiKeyValue + apiKeyValue);
+    private Authentication createAdminsAuthToken(String apiKeyValue) {
+        log.debug("Creating admins authentication token using API key with the value {}.", apiKeyValue);
+        Set<SimpleGrantedAuthority> authoritiesGrantedToYggAdmin = new HashSet<>();
+        authoritiesGrantedToYggAdmin.add(new SimpleGrantedAuthority(
+                ServiceApiAccessRightsVoter.ACCESS_SERVICE_AUTHORITY_PREFIX + "refresh"));
+        ApiKeyAuthToken apiKeyAuthToken = new ApiKeyAuthToken(apiKeyValue, authoritiesGrantedToYggAdmin);
+        apiKeyAuthToken.setAuthenticated(true);
+        log.info("Admin is successfully authenticated using the API key with the value {}.", apiKeyValue);
+        return apiKeyAuthToken;
+    }
 
-            ResponseEntity<ServiceApi> responseEntity = restTemplate.getForEntity(urlToGetServiceApi, ServiceApi.class);
+    private Authentication verifyApiKeyAndIfAuthorizedCreateConsumerAuthToken(String apiKeyValue) {
+        log.debug("Verifying the API with the value {} from consumer.", apiKeyValue);
 
-            if (responseEntity != null) {
-                ServiceApi serviceApi = responseEntity.getBody();
-                log.debug("Service API was found by CoatRack admin: " + serviceApi);
+        boolean isApiKeyValid;
+        ApiKey apiKey;
 
-                return serviceApi;
-            } else {
-                log.error("Communication with Admin server failed, result is: " + responseEntity);
-                throw new AuthenticationServiceException("Communication with auth server failed");
-            }
-        } catch (Exception e) {
-            throw new AuthenticationServiceException("An error occured, when trying to get service API data from admin server", e);
+        try {
+            apiKey = apiKeyFetcher.requestApiKeyFromAdmin(apiKeyValue);
+            isApiKeyValid = localApiKeyVerifier.isApiKeyValid(apiKey);
+        } catch (RestClientException e) {
```

Comment:
Strictly speaking, this method should not need to know at all that there is a rest client involved. If there is a rest client exception, that should be catched and handled in the `apiKeyFetcher`.

Reply:
This is now solved by introducing a ApiKeyFetchingException.